### PR TITLE
Fix `./bin` being in `$PATH`

### PR DESCRIPTION
**Pull request by ChillerDragon**
# since when

Introduced in this commit:
https://github.com/Homebrew/install/commit/ce3f3e050e6afa4a9876a4d07fda889310dff977 @gromgit (you probably care about this)

# reproduce

To reproduce a failing install make sure that ./bin is in your path

```
export PATH="./bin:$PATH"
```

Then run the installer.

```
/bin/bash -c "$(curl -fsSL
https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

It will fail like this

```
==> Downloading and installing Homebrew...
/bin/bash: line 234: ./bin/git: No such file or directory
Failed during: ./bin/git -c init.defaultBranch=master init --quiet
```

# why does it fail

Because when the installer checks for a git executable it is currently in the /usr directory.
See
https://github.com/Homebrew/install/blob/95648ef45c8d59a44fa4ab8f29cdcf17d6ec48ac/install.sh#L447

And in the /usr directory there is /usr/bin/git so the ``command -v``will find ``./bin/git`` and try to use that as git executable. See
https://github.com/Homebrew/install/blob/95648ef45c8d59a44fa4ab8f29cdcf17d6ec48ac/install.sh#L450

One could argue that having ./bin in PATH is a broken setup but its common to prio local binaries on a per project/folder basis. And I guess the whole point of the installer is to make it work on every system.

tl;dr

```
$ export PATH="./bin:$PATH"
$ cd /usr/
$ command -v git
./bin/git

```

# the fix

One could ensure that never a relative PATH is used by prefixing ``$USABLE_GIT`` with ``pwd`` but im not sure if that covers all edge cases and is the cleanest solution. I noticed that `command -V` prints the full path at least on my system (debian). Not sure how standard that is could some mac user verify that the fix works please?

The fix i went for is command -V instead of command -v

 ```
$ command -v git
./bin/git
$ command -V git
git is /usr/bin/git
$ command -V git | awk '{ print $3 }'
/usr/bin/git
 ```

### CODE REVIEW

**Comment by SMillerDev** [comment]
Putting relative locations in `$PATH` just seems like a recipe for disaster either way, I'm not sure the installer should protect against that

**Comment by vitorgalvao** [comment]
> I'm not sure the installer should protect against that

Agreed. Plus, I’m not sure this would work:

```console
$ command -V fakegit | awk '{ print $3 }'
found
```

Because without `awk` it returns `fakegit not found`. So if `git` isn’t there this check would still succeed despite being a failure.

**Comment by ChillerDragon** [comment]
> Plus, I’m not sure this PR would work.
> 
> ```
> $ command -V fakegit | awk '{ print $3 }'
> found
> ```
> 
> Because without `awk` it returns `fakegit not found`. So if `git` isn’t there this check would still succeed despite being a failure.

Interesting on my system it prints to stderr thus sets the variable to be empty

```
$ USABLE_GIT="$(command -V fakegit | awk '{ print $3 }')"
-bash: command: fakegit: not found
$ echo $USABLE_GIT

$ 
```

But yea agree the fix is for sure flawed if it does not work on your system. But some fix would be nice or at least an warning if a relative path is matched? Such as:

```
[[ "${USABLE_GIT::1}" == "." ]] && echo "Warning: relative git path found please remove relatives paths from your PATH"
```

This installer script is run on a lot of systems and having ./bin in the PATH might be bad practice but it is a thing. Just crashing the installer seems a bit intense.

**Comment by gromgit** [comment]
Sorry, I think this PR is heading in the wrong direction. What `install.sh` needs is a known-good `git` that can be executed from any directory. Therefore, all that's needed is to fully-qualify `${USABLE_GIT}` before using it.

Come to think of it, the same should be done for `curl` and `${USABLE_CURL}`, so this entire `USABLE_xxx` code section bears refactoring.

Also, just noticed that `install.sh` on macOS blindly uses the first `git` seen. That doesn't seem right. Does anyone remember why it was done this way?

File: install.sh [hunk 1]
```diff
@@ -447,8 +447,8 @@ fi
 cd "/usr" || exit 1
 
 ####################################################################### script
-USABLE_GIT="$(command -v git)"
-if [[ -z "${USABLE_GIT}" ]]
+USABLE_GIT="$(command -V git | awk '{ print $3 }')"
```
Comment (carlocab) [inline install.sh:450]: Why don't we just check if
```
command -v git
```
starts with `.`, and if so, replace the `.` with the current working directory?

Comment (gromgit) [reply install.sh:450]: Note that relative path components in `PATH` don't have to start with a `.`: `PATH=mybin:$PATH command -v git` can return `mybin/git`.

Better to `readlink -f` the result instead, that always Does The Right Thing.

Comment (ChillerDragon) [reply install.sh:450]: Yea as mentioned in the pr this appending was a thought of mine. But checking the first character seemed like adding a lot of complexity. What would work is checking if it is a / but then there seems to be so much that could go wrong.

I also thought about ``readlink`` but is that a standard preinstalled tool on all systems?

**Comment by MikeMcQuaid** [comment]
> Sorry, I think this PR is heading in the wrong direction. What install.sh needs is a known-good git that can be executed from any directory. Therefore, what's really needed is to fully-qualify ${USABLE_GIT} before using it.
> Come to think of it, the same should be done for curl and ${USABLE_CURL}, so this entire USABLE_xxx code section bears refactoring.

Agreed.

> Also, just noticed that install.sh on macOS blindly uses the first git (and curl) seen. That doesn't seem right. Does anyone remember why it was done this way?

Because this all works fine. The system versions are all we want to use on macOS, really, on all supported versions.

File: install.sh (hunk 1 above)
Comment (gromgit) [reply install.sh:450]: > I also thought about `readlink` but is that a standard preinstalled tool on all systems?

`readlink` is standard fare on macOS, going back at least to Mojave.

As for Linux, given that `readlink` is a part of GNU `coreutils`, I'm pretty sure any distro missing `readlink` is also missing enough other essential utilities that we actually _want_ the installer to fail, noisily.

Comment (MikeMcQuaid) [reply install.sh:450]: Could try `realpath` if it's only for Linux anyway?

Comment (MikeMcQuaid) [reply install.sh:450]: Alternatively: https://stackoverflow.com/a/31605674 (the `cd` thing) is the standard one

**Comment by ChillerDragon** [comment]
Thanks @gromgit for the [note about](https://github.com/Homebrew/install/pull/752#discussion_r1124527248) readlink being portable. I force pushed the fix to use readlink instead.

Works fine on my system even with git not being found but would be nice if some mac users could verify that nothing weird happens.

```
/usr$ export PATH="./bin:$PATH"
/usr$ command -v git
./bin/git
/usr$ USABLE_GIT="$(readlink -f "$(command -v git)")"
/usr$ echo $USABLE_GIT 
/usr/bin/git
/usr$ USABLE_GIT="$(readlink -f "$(command -v fakegit)")"                                                                                                                                             
/usr$ echo $USABLE_GIT

/usr$ 
```

File: install.sh (hunk 1 above)
Comment (ChillerDragon) [reply install.sh]: > Could try `realpath` if it's only for Linux anyway?

The bug this pr tries to fix exists on mac and linux at the moment.

File: install.sh [hunk 2]
```diff
@@ -447,7 +447,7 @@ fi
 cd "/usr" || exit 1
 
 ####################################################################### script
-USABLE_GIT="$(command -v git)"
+USABLE_GIT="$(readlink -f "$(command -v git)")"
```
Comment (MikeMcQuaid) [inline install.sh:450]: Need to only use `readlink` here if it's a symlink otherwise things go pretty weird:

```
$ readlink -f $(/usr/bin/git)
/opt/homebrew/etc
```

Comment (gromgit) [reply install.sh:450]: >```
>readlink -f $(/usr/bin/git)
>```

Doesn't that try to parse the _output_ of `/usr/bin/git` as a pathname?

Comment (MikeMcQuaid) [reply install.sh:450]: It does. I am apparently a moron 😂.

Ignore me, good to ship!

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution (hopefully of many)! Without people like you submitting PRs we couldn't run this project. You rock, @ChillerDragon!

**Comment by gromgit** [comment]
> `readlink` is standard fare on macOS, going back at least to Mojave.

My bad, I was looking at (and testing on) M1 Monterey when I made that statement. While `readlink` has been in macOS for a long while now, the `-f` flag is new to Monterey. :man_facepalming:

**Comment by MikeMcQuaid** [comment]
Sorry @ChillerDragon, had to revert this due to `readlink -f` being missing on older versions as @gromgit mentioned.

**Comment by ChillerDragon** [comment]
> Sorry @ChillerDragon, had to revert this due to `readlink -f` being missing on older versions as @gromgit mentioned.

Woopsie. Okay what about going with the fix I force pushed away in favor of readlink.
Or is the ``-V`` also not portable?

```bash
USABLE_GIT="$(command -V fakegit | awk '{ print $3 }')"
if [[ -z "${USABLE_GIT}" ]] && [[ "${USABLE_GIT}" != found ]]
```

File: install.sh (hunk 1 above)
Comment (vitorgalvao) [reply install.sh]: > Could try `realpath` if it's only for Linux anyway?

Worth noting `/bin/realpath` is now a thing on Ventura. So in a few years it would be a viable solution for Homebrew on macOS too.

**Comment by vitorgalvao** [comment]
> we could add this as a dependency https://github.com/ko1nksm/readlinkf

I would say that‘s definitely too much.

> Also `realpath ./bin/git` works the same as `readlink -f ./bin/git` on my system

`realpath` is only available from Ventura, so it’s worse.

**Comment by SMillerDev** [comment]
I think a much better solution is if the installer just screamed really loud if it found relative paths in `PATH` since that's a recipe for disaster whichever way you look at it and then the user can fix it for themselves.

**Comment by ChillerDragon** [comment]
> I think a much better solution is if the installer just screamed really loud if it found relative paths in `PATH` since that's a recipe for disaster whichever way you look at it and then the user can fix it for themselves.

You mean like adding something like that at the top of the installer?

```bash
if echo "$PATH" | grep -qEo '(^|:)[^/][^:]+'
then

	echo "$PATH" | grep -Eo '(^|:)[^/][^:]+'
	abort 'Can not proceed with those relatives paths in $PATH'
fi
```

It is for sure better than the current state that fails with this output:

```
/bin/bash: line 234: ./bin/git: No such file or directory
Failed during: ./bin/git -c init.defaultBranch=master init --quiet
```

But I would strongly advise against breaking the installer for every user with this common setup. You might have a point there that this setup has its flaws but the reality is that most shell rc files are flawed.

**Comment by gromgit** [comment]
I'm thinking more like the following, to balance "informing the user when something's not right" with "making sure we get a tool we can use":
```
diff --git a/install.sh b/install.sh
index ca4f7ba..a7f9815 100755
--- a/install.sh
+++ b/install.sh
@@ -405,7 +405,10 @@ find_tool() {
   local executable
   while read -r executable
   do
-    if "test_$1" "${executable}"
+    if [[ "$executable" != /* ]]
+    then
+      warn "Ignoring $executable - relative paths don't work"
+    elif "test_$1" "${executable}"
     then
       echo "${executable}"
       break
```
